### PR TITLE
Move NonLocalTOperator objects from Hamiltonian to Drivers

### PR DESCRIPTION
**Pull request by ye-luo**
Review after #5135
## Proposed changes
1. NonLocalTOperator object is no longer one per Hamiltonian but per QMCUpdateBase or crowd.
2. Hamiltonian no longer needs to keep track of nlpp_ptr which only works when there is only one NLPP. 
3. Make makeNonLocalMovesPbyP virtual and SO NLPP can implement it later
4. NonLocalTOperator class still needs some cleanup around constructor which I keep it intact.
5. DMCContextForSteps inherited from ContextForSteps to include a NonLocalTOperator object.

## What type(s) of changes does this code introduce?
- New feature

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
laptop

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted

### CODE REVIEW

File: src/QMCDrivers/QMCDriverNew.h [hunk 1]
```diff
@@ -112,6 +111,16 @@ class QMCDriverNew : public QMCDriverInterface, public MPIObjectBase
   //xmlNodePtr walker_logs_xml;
 
 protected:
+  class ContextForSteps
+  {
```
Comment (prckent) [inline src/QMCDrivers/QMCDriverNew.h:115]: Can you add a description?

File: src/QMCDrivers/DMC/DMCBatched.cpp [hunk 2]
```diff
@@ -33,13 +33,25 @@
...
 
+class DMCBatched::DMCContextForSteps : public ContextForSteps
+{
+public:
+  DMCContextForSteps(RandomBase<FullPrecRealType>& random_gen, const NonLocalTOperator& non_local_ops)
+      : ContextForSteps(random_gen), non_local_ops(non_local_ops)
+  {}
+
+  ///non local operator
+  NonLocalTOperator non_local_ops;
```
Comment (prckent) [inline src/QMCDrivers/DMC/DMCBatched.cpp:52]: Can this be protected or private?

**Review by prckent — CHANGES_REQUESTED** [review]

File: src/QMCDrivers/QMCDriverNew.h (hunk 1 above)
Comment (ye-luo) [reply src/QMCDrivers/QMCDriverNew.h:115]: will do.

File: src/QMCDrivers/DMC/DMCBatched.cpp (hunk 2 above)
Comment (ye-luo) [reply src/QMCDrivers/DMC/DMCBatched.cpp:52]: ContextForSteps and its derived class are intended to be just a struct. Marking it private and adding an access function are unnecessary.

File: src/QMCDrivers/QMCDriverNew.h (hunk 1 above)
Comment (ye-luo) [reply src/QMCDrivers/QMCDriverNew.h:116]: fixed.

**Comment by ye-luo** [comment]
Test this please

**Review by prckent — APPROVED** [review]

**Pull request merged**